libs/stdlib/iec61131-st/to_string.st

```diff
@@ -0,0 +1,125 @@
+FUNCTION TO_CHAR<T: ANY> : CHAR
+    VAR_INPUT
+        in : T;
+    END_VAR
+END_FUNCTION
+
+FUNCTION TO_CHAR__CHAR : CHAR
+    VAR_INPUT
+        in : CHAR;
+    END_VAR
+
+    TO_CHAR__CHAR := in;
+END_FUNCTION
+
+FUNCTION TO_CHAR__WCHAR : CHAR
+    VAR_INPUT
+        in : WCHAR;
+    END_VAR
+
+    TO_CHAR__WCHAR := WCHAR_TO_CHAR(in);
+END_FUNCTION
+
+FUNCTION TO_CHAR__STRING : CHAR
+    VAR_INPUT {ref}
```

Comment:
The `{ref}` here is no longer necessary, since aggregate-type arguments are now always passed as pointers. Not blocking, we should probably change these for all occurances in the `stdlib

Reply:
Let's address this in another PR since I also forgot to remove the code-gen python scripts.

Reply:
I am confused, why would Aggregate types be implicitly pass by reference that isn't how it works in Codesys. they are passed by value unless you explicitly pass by reference or use a Reference to or Pointer type. 


Maybe I am.missing somthing.

Reply:
This would be an implicit reference that the user would not see. I don't know the codesys internal representation but I strongly suspect they would also do something similar for chars and string.

There are multiple reasons to do this but mainly it's because that's what C and llvm usually expect but also compile time performance since the alternative would cause us to preallocate values on the call stack and we previously had compile issues with that. The effect of this is as mentioned transparent for the user in PLC. The C function would however have an additional parameter.

What we do internally to preserve the by value behavior is to actually allocate a new variable on the stack and pass a pointer to that variable

Reply:
Oh ok, so are you saying there would be a new variables allocated on the heap, and then the value would be copied in to it? Ok I think I am following how that makes sense.

Thank you for the explanation.

Reply:
Still on the stack, just allocated at runtime because allocating/returning at compile time was very slow during the optimisation passes / binary generation pass
https://llvm.org/docs/LangRef.html#alloca-instruction

Reply:
Here's some more context on why this is usually needed: 
https://stackoverflow.com/questions/11656532/returning-an-array-using-c
Technically it's not required for structs but it made our lives easier there. 

This is an equivalant C application with a struct and a string allocated on the stack using this method:
```c
#include<stdio.h>

typedef struct {
	int x;
	int y;
} MyStruct;


void func(MyStruct* str) {
	str->x = 1;
	str->y = 2;
}

void stringExample(char* arg) {
	sprintf(arg, "Hello, World!");
}

int main() {
	MyStruct str;
	printf("Before: x : %d y: %d\n", str.x, str.y);
	func(&str);
	printf("After: x : %d y: %d\n", str.x, str.y);
	char helloWorld[14];
	stringExample(helloWorld);
	printf("%s\n", helloWorld);
	return 0;
}

```

Reply:
Also for context, I think the graph speaks for itself but the big reduction in compile times (oscat) was due to that internal change https://plc-lang.github.io/metrics/

Reply:
> Here's some more context on why this is usually needed: 
> https://stackoverflow.com/questions/11656532/returning-an-array-using-c
> Technically it's not required for structs but it made our lives easier there. 
> 
> Edit: I got something confused here, I gave an example with returns in mind, but these are parameters, the same issues still apply.. we do indeed copy the values before passing them using a memcpy but it's still all on the stack.
> https://llvm.org/docs/LangRef.html#llvm-memcpy-intrinsic
> 
> This is an equivalant C application with a struct and a string allocated on the stack using this method:
> ```c
> #include<stdio.h>
> 
> typedef struct {
> 	int x;
> 	int y;
> } MyStruct;
> 
> 
> void func(MyStruct* str) {
> 	str->x = 1;
> 	str->y = 2;
> }
> 
> void stringExample(char* arg) {
> 	sprintf(arg, "Hello, World!");
> }
> 
> int main() {
> 	MyStruct str;
> 	printf("Before: x : %d y: %d\n", str.x, str.y);
> 	func(&str);
> 	printf("After: x : %d y: %d\n", str.x, str.y);
> 	char helloWorld[14];
> 	stringExample(helloWorld);
> 	printf("%s\n", helloWorld);
> 	return 0;
> }
> 
> ```

Ooooh ok, I didn't known dynamically allocating to the stack was even possible. I guess it makes sense that it is. 

I get it now, it makes sense. 

This wiki article helped it finally click in my brain. 
https://en.m.wikipedia.org/wiki/Stack-based_memory_allocation#:~:text=Many%20Unix%2Dlike%20systems%20as,variable%2Dlength%20arrays%20are%20handled.

Reply:
> Here's some more context on why this is usually needed: 
> https://stackoverflow.com/questions/11656532/returning-an-array-using-c
> Technically it's not required for structs but it made our lives easier there. 
> 
> Edit: I got something confused here, I gave an example with returns in mind, but these are parameters, the same issues still apply.. we do indeed copy the values before passing them using a memcpy but it's still all on the stack.
> https://llvm.org/docs/LangRef.html#llvm-memcpy-intrinsic
> 
> This is an equivalant C application with a struct and a string allocated on the stack using this method:
> ```c
> #include<stdio.h>
> 
> typedef struct {
> 	int x;
> 	int y;
> } MyStruct;
> 
> 
> void func(MyStruct* str) {
> 	str->x = 1;
> 	str->y = 2;
> }
> 
> void stringExample(char* arg) {
> 	sprintf(arg, "Hello, World!");
> }
> 
> int main() {
> 	MyStruct str;
> 	printf("Before: x : %d y: %d\n", str.x, str.y);
> 	func(&str);
> 	printf("After: x : %d y: %d\n", str.x, str.y);
> 	char helloWorld[14];
> 	stringExample(helloWorld);
> 	printf("%s\n", helloWorld);
> 	return 0;
> }
> 
> ```

Ooooh I understand now. I never really consider being able to dynamically allocated to the stack. So I am following you now. 

Tha know you for the explanation!!! Veey much appreciated sorry to derail your PR discussion.